### PR TITLE
Implemented pixel format transcoding for Bitmap and WriteableBitmap

### PR DESCRIPTION
**Pull request by kekekeks**
- PixelFormat is now an opaque struct
- Implemented support for bitmap data with pixel formats not supported by Skia/Direct2D1 (for now various grayscale formats)
- Implemented Bitmap.CopyPixels

### CODE REVIEW

File: src/Avalonia.Base/Platform/PixelFormat.cs [hunk 1]
```diff
@@ -1,9 +1,99 @@
...
+        BlackWhite,
+        Gray2,
+        Gray4,
+        Gray8,
+        Gray16,
+        Gray32Float,
+        Rgba64
+    }
+
+    public struct PixelFormat : IEquatable<PixelFormat>
```
Comment (maxkatz6) [inline src/Avalonia.Base/Platform/PixelFormat.cs:19]: Can be a "recrod struct" without IEquatable/Equals and stuff
https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTACX3YAuApgE4wQA22RMArmNgApQAeRFAYgPYmQECidMDjwBvEXjwAlAObAArADZ5AGgmTss4BAAcenWowa8AIRkld+w8dwmKEAMYBrAOoALKMWs2A4hYCeSN7GfhD+ACzBGqH+BuqSMSiKUQkBAMxIHBRcEAQp0nIQiuHqAL7q6ghp2CREDjwAJtgAzgQktA4ETKzs3Ly56uJGNoSk5FTMbJw8fIL02H2zQgDcmPF4VfgwXSaezYykk+zruEM2GjJEBCcaZ+cjAGbYABSLuXMMALyf3VNvAkIAHR2RyuDzEACUN3uGwA7NgUKthjDJOxmkR8E9XjN3kJsN9fr0cQD6ICYkgociUcYEPCkEjqaiKOjMS9/h98T8jtN+iSwGSAuFKYybLTsOEGSLqMyMVAsey8QTuQrSTEdMKpZIxTpJSK0bL5cSOUqejylqStEp5NhoZq4HAFkbFVzTSr+YlFBrNbgxUldYz9azsbzjS6/k6LYVil7vWLikjbTT4Rl/edylTJOmbqMyJRsMAuFwqAAJCDNACCFAADm4IPiAHyOkPOwlm3GR7T6HQ2jPe6n2pvmr5honN0lmCxdxN9jQDt2c1tuwFaIoSta93A58aL4nPZURhgPYkxyR3GHzn5H3mpvBZjfYAD0D82CCSO95mjkVob774zWXX7KDeuBPi+b77h+K5dj+EF/gBnb6MBj7PtUr6KL+uTYBOlh6DBrrEv+2Fdkh67nJsXAAG6kCQUANBir4AAzYAAKlwADKbSwDIzwQj+S6sRxNEwNxEL+qUQA==

Either way it's still can be created from the user code with public ctor

File: src/Avalonia.Base/Platform/IBitmapWithPixelReadAccessImpl.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+namespace Avalonia.Platform;
+
+public interface IReadableBitmapImpl
```
Comment (maxkatz6) [inline src/Avalonia.Base/Platform/IBitmapWithPixelReadAccessImpl.cs:3]: File name and interface names are not matching.

Comment (maxkatz6) [reply src/Avalonia.Base/Platform/IBitmapWithPixelReadAccessImpl.cs:3]: Also doesn't `ILockedFramebuffer Lock()` provide write access to the bitmap as well?

File: src/Avalonia.Base/Media/Imaging/PixelFormatReaders.cs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+        else if (format == PixelFormats.Gray8)
+            Transcode<Gray8PixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Gray16)
+            Transcode<Gray16PixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Gray32Float)
+            Transcode<Gray32FloatPixelReader>(dst, src, size, strideSrc, strideDst);
+        else if (format == PixelFormats.Rgba64)
+            Transcode<Rgba64PixelFormatReader>(dst, src, size, strideSrc, strideDst);
+        else
+            throw new NotSupportedException($"Pixel format {format} is not supported");
```
Comment (maxkatz6) [inline src/Avalonia.Base/Media/Imaging/PixelFormatReaders.cs:250]: From the list of these formats, seems like Rgb565 is missing as D2D backend doesn't support it (as I see from this PR). Low priority, I suppose.

**Review by maxkatz6 — APPROVED** [review]

File: src/Avalonia.Base/Platform/IBitmapWithPixelReadAccessImpl.cs (hunk 2 above)
Comment (kekekeks) [reply src/Avalonia.Base/Platform/IBitmapWithPixelReadAccessImpl.cs]: It does, but it's `Impl`-suffixed interface that's not a public API.

In general WriteableBitmap should have `AddDirtyRect` API to notify Avalonia that something got changed.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029207-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**